common/flatpak-run-wayland.c

```diff
@@ -217,5 +292,17 @@ flatpak_run_add_wayland_args (FlatpakBwrap *bwrap,
                               NULL);
       flatpak_bwrap_add_runtime_dir_member (bwrap, wayland_display);
     }
+
+  if (!inherit_wayland_socket)
+    {
+      flatpak_bwrap_unset_env (bwrap, "WAYALND_SOCKET");
+      /* In order avoid thread-safety issue of setenv/unsetenv, set FD_CLOEXEC on
+       * it instead of close it so the fd won't be leak to application. Otherwise
+       * file descriptor reuse and multi threading may potentially make us close
+       * the wrong file descriptor. */
+      fd = get_wayland_socket_fd ();
+      if (fd >= 0)
+        fcntl(fd, F_SETFD, FD_CLOEXEC);
```

Comment:
I don't buy the explanation here. There is no code modifying the environment variable and if there was some race, connecting would be equally broken. I think we can just close the socket. Let's get feedback from @smcv as well.

Reply:
If this function runs twice, the second time wayland_socket will still be the old value. If you close the fd in a previous run, you may close a totally wrong fd in the second run if another fd is opened later.

Unless we unset it or use some other global to hold the value. But unset it will cause the thread safety issue.

Reply:
I think @wengxt does have a valid point here. We're working with two environment blocks:

1. The process-global `environ`, which we must not modify
2. The copy that we will use to invoke bubblewrap, `bwrap->envp`, which we can modify in the thread that owns the `FlatpakBwrap` object (our main thread)

Let's suppose we inherited the private Wayland socket as file descriptor 5, so we have `WAYLAND_SOCKET=5` in `environ`, and we have also copied that into `bwrap->envp`.

If I understand correctly, what @wengxt initially wanted to do was the equivalent of this pseudocode, but with better error handling:

```
if (getenv ("WAYLAND_SOCKET"))
  close (atoi (getenv ("WAYLAND_SOCKET")));    /* closes it if still present in environ */

unsetenv ("WAYLAND_SOCKET");    /* removes it from environ */
```

But we can't do that, because as previously discussed, we can't safely modify `environ` while a background thread might be doing a `getenv()`.

So, I think the next thing @wengxt considered doing was this:

```
remove "WAYLAND_SOCKET" from bwrap->envp;

if (getenv ("WAYLAND_SOCKET"))
  close (atoi (getenv ("WAYLAND_SOCKET")));
```

But @wengxt is correct to say that if we somehow did that twice, it would be unsafe: the first time, we would close the fd 5 that we inherited from the caller, and the second time, we could potentially close an unrelated fd 5 that had been opened in the meantime by Flatpak code. So that would be wrong to do.

In reality, I don't think we ever call this function twice, but it seems fragile to rely on that.

What I think @wengxt might be missing here is that when `flatpak run` invokes bubblewrap, it automatically makes *every* file descriptor close-on-exec, except for the ones that have specifically been marked as inheritable. (This is done in `flatpak_bwrap_child_setup()`.) So, strictly speaking, making the fd close-on-exec in this function is unnecessary: it will be close-on-exec anyway, unless we add it to `bwrap->fds`, which we don't do in this code path.

Reply:
use flatpak_bwrap_add_fd to attach the fd.